{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Notebook\n",
    "\n",
    "This my first experimtn\n",
    "\n",
    "## Reference\n",
    "\n",
    " \n",
    "\n",
    "## DataSets\n",
    "\n",
    " * Mouse Genome: mm10 ftp://igenome:G3nom3s4u@ussd-ftp.illumina.com/Mus_musculus/UCSC/mm10/Mus_musculus_UCSC_mm10.tar.gz\n",
    " \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Set the environment variables to be used during the project \n",
    "\n",
    "import os\n",
    "import xmltodict\n",
    "from Bio import Entrez\n",
    "Entrez.email = \"r78v10a07@gmail.com\"\n",
    "os.environ['WORKDIR'] = os.path.abspath('../')\n",
    "os.environ['CONFIG'] = os.environ['WORKDIR'] + '/config'\n",
    "os.environ['DATA'] = os.environ['WORKDIR'] + '/data'\n",
    "os.environ['BIN'] = os.environ['WORKDIR'] + '/bin'\n",
    "os.environ['RESULTS'] = os.environ['WORKDIR'] + '/results'\n",
    "os.environ['DOC'] = os.environ['WORKDIR'] + '/docs'\n",
    "os.environ['SRC'] = os.environ['WORKDIR'] + '/src'\n",
    "\n",
    "os.environ['MOUSE_BIOPROJECT'] = 'PRJNA79859'\n",
    "os.environ['MOUSE_SRA'] = 'SRA026846'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# Downloading the mm10 genome to the folder data/mm10\n",
    "cd $DATA\n",
    "if [ ! -e \"Mus_musculus\" ]\n",
    "then    \n",
    "    wget -b ftp://igenome:G3nom3s4u@ussd-ftp.illumina.com/Mus_musculus/UCSC/mm10/Mus_musculus_UCSC_mm10.tar.gz\n",
    "    tar xzf Mus_musculus_UCSC_mm10.tar.gz\n",
    "fi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Get the SRA IDs related to the BioProject from NCBI\n",
    "\n",
    "handle = Entrez.esearch(db=\"sra\", term=os.environ['MOUSE_BIOPROJECT'], retmax=50)\n",
    "mouse_sra_id = Entrez.read(handle)\n",
    "handle.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# Retrieving the bioproject from NCBI\n",
    "\n",
    "handle = Entrez.esearch(db=\"bioproject\", term=os.environ['MOUSE_BIOPROJECT'])\n",
    "mouse_bioproject_id = Entrez.read(handle)\n",
    "handle.close()\n",
    "mouse_bioproject_id = mouse_bioproject_id['IdList']\n",
    "handle = Entrez.efetch(db=\"bioproject\", id=mouse_bioproject_id)\n",
    "mouse_bioproject = xmltodict.parse(handle.read())\n",
    "handle.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Retirve each experiment related to the SRA IDs from NCBI\n",
    "\n",
    "mouse_sra = {}\n",
    "for id in mouse_sra_id['IdList']:\n",
    "    handle = Entrez.efetch(db=\"sra\", id=id, rettype=\"full\", retmode=\"full\")\n",
    "    mouse_sra[id] = xmltodict.parse(handle.read())\n",
    "    handle.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID:\t40432\tTITLE:\tB6 lane 1 experiment 7\tALIAS:\tB6_1_7\tRUN:\tSRR099237\n",
      "ID:\t40420\tTITLE:\tB6 lane 6 experiment 4\tALIAS:\tB6_6_4\tRUN:\tSRR099227\n",
      "ID:\t40429\tTITLE:\tD2 lane 7 experiment 6\tALIAS:\tD2_7_6\tRUN:\tSRR099235\n",
      "ID:\t40423\tTITLE:\tB6 lane 8 experiment 4\tALIAS:\tB6_8_4\tRUN:\tSRR099229\n",
      "ID:\t40428\tTITLE:\tD2 lane 6 experiment 6\tALIAS:\tD2_6_6\tRUN:\tSRR099234\n",
      "ID:\t40435\tTITLE:\tD2 lane 5 experiment 7\tALIAS:\tD2_5_7\tRUN:\tSRR099240\n",
      "ID:\t40424\tTITLE:\tB6 lane 1 experiment 6\tALIAS:\tB6_1_6\tRUN:\tSRR099230\n",
      "ID:\t40419\tTITLE:\tD2 lane 5 experiment 4\tALIAS:\tD2_5_4\tRUN:\tSRR099226\n",
      "ID:\t40427\tTITLE:\tB6 lane 5 experiment 6\tALIAS:\tB6_5_6\tRUN:\tSRR099233\n",
      "ID:\t40430\tTITLE:\tD2 lane 8 experiment 6\tALIAS:\tD2_8_6\tRUN:\tSRR099236\n",
      "ID:\t40426\tTITLE:\tB6 lane 3 experiment 6\tALIAS:\tB6_3_6\tRUN:\tSRR099232\n",
      "ID:\t40437\tTITLE:\tD2 lane 7 experiment 7\tALIAS:\tD2_7_7\tRUN:\tSRR099242\n",
      "ID:\t40417\tTITLE:\tD2 lane 2 experiment 4\tALIAS:\tD2_2_4\tRUN:\tSRR099224\n",
      "ID:\t40416\tTITLE:\tD2 lane 1 experiment 4\tALIAS:\tD2_1_4\tRUN:\tSRR099223\n",
      "ID:\t40438\tTITLE:\tD2 lane 8 experiment 7\tALIAS:\tD2_8_7\tRUN:\tSRR099243\n",
      "ID:\t40433\tTITLE:\tB6 lane 2 experiment 7\tALIAS:\tB6_2_7\tRUN:\tSRR099238\n",
      "ID:\t40436\tTITLE:\tD2 lane 6 experiment 7\tALIAS:\tD2_6_7\tRUN:\tSRR099241\n",
      "ID:\t40425\tTITLE:\tB6 lane 2 experiment 6\tALIAS:\tB6_2_6\tRUN:\tSRR099231\n",
      "ID:\t40422\tTITLE:\tB6 lane 7 experiment 4\tALIAS:\tB6_7_4\tRUN:\tSRR099228\n",
      "ID:\t40434\tTITLE:\tB6 lane 3 experiment 7\tALIAS:\tB6_3_7\tRUN:\tSRR099239\n",
      "ID:\t40418\tTITLE:\tD2 lane 3 experiment 4\tALIAS:\tD2_3_4\tRUN:\tSRR099225\n"
     ]
    }
   ],
   "source": [
    "# Let's print the runs retrieved\n",
    "\n",
    "for rec in mouse_sra:\n",
    "    print('ID:\\t' + rec \n",
    "          + '\\tTITLE:\\t' + mouse_sra[rec]['EXPERIMENT_PACKAGE_SET']['EXPERIMENT_PACKAGE']['SAMPLE']['TITLE']\n",
    "          + '\\tALIAS:\\t' + mouse_sra[rec]['EXPERIMENT_PACKAGE_SET']['EXPERIMENT_PACKAGE']['SAMPLE']['@alias']\n",
    "          + '\\tRUN:\\t' + mouse_sra[rec]['EXPERIMENT_PACKAGE_SET']['EXPERIMENT_PACKAGE']['RUN_SET']['RUN']['@accession'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The data is ready\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "# Downloading the SRR data using fastq-dump\n",
    "\n",
    "if [ ! -e \"$DATA/${MOUSE_SRA}\" ]\n",
    "then\n",
    "    mkdir $DATA/${MOUSE_SRA}\n",
    "    sh $BIN/download_SRA.sh ${MOUSE_SRA} $DATA/${MOUSE_SRA}\n",
    "else\n",
    "    echo \"The data is ready\"\n",
    "fi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running bowtie2 for SRR099228\n",
      "Running bowtie2 for SRR099243\n",
      "Running bowtie2 for SRR099236\n",
      "Running bowtie2 for SRR099237\n",
      "Running bowtie2 for SRR099233\n",
      "Running bowtie2 for SRR099226\n",
      "Running bowtie2 for SRR099235\n",
      "Running bowtie2 for SRR099240\n",
      "Running bowtie2 for SRR099242\n",
      "Running bowtie2 for SRR099231\n",
      "Running bowtie2 for SRR099238\n",
      "Running bowtie2 for SRR099225\n",
      "Running bowtie2 for SRR099241\n",
      "Running bowtie2 for SRR099227\n",
      "Running bowtie2 for SRR099239\n",
      "Running bowtie2 for SRR099224\n",
      "Running bowtie2 for SRR099229\n",
      "Running bowtie2 for SRR099234\n",
      "Running bowtie2 for SRR099232\n",
      "Running bowtie2 for SRR099230\n",
      "Running bowtie2 for SRR099223\n"
     ]
    }
   ],
   "source": [
    "# Let's run Bowtie2 for each SRR\n",
    "\n",
    "os.chdir(os.environ['RESULTS'])\n",
    "for rec in mouse_sra:\n",
    "    os.environ['SAMPLE'] = mouse_sra[rec]['EXPERIMENT_PACKAGE_SET']['EXPERIMENT_PACKAGE']['RUN_SET']['RUN']['@accession']\n",
    "    print(\"Running bowtie2 for \" + os.environ['SAMPLE'])\n",
    "    !bowtie2 -x ../data/Mus_musculus/UCSC/mm10/Sequence/Bowtie2Index/genome --very-sensitive --score-min \"L,-0.4,-0.1\" --gbar 8 -p 4 $DATA/${MOUSE_SRA}/$SAMPLE.fastq.gz | samtools view -Sb - > $SAMPLE.bam\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.4.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
